### DISCUSSION TITLE
Stache - Python forked version by clach04

### DISCUSSION THREAD

**Discussion by clach04** [mustache/spec#172] [Show and tell]
There are a few different Mustache implementations for Python (including HandleBars). If you are looking for a Python library I'd probably recommend:

1. https://github.com/PennyDreadfulMTG/pystache - which is actively maintained as of 2023-08-26
      * Python 3.x only
      * Available from PyPi (via pip)
2. https://github.com/noahmorrison/chevron - not currently active (but there are some active forks)
      * Python 2.6, 2.7, and 3.x
      * claims to follow Mustache spec closely (with test suite)
      * claims to be the fastest implementation (with a benchmark available)

My hard fork is https://github.com/clach04/Stache as the upstream one (https://github.com/hyperturtle/Stache) is not active and pull requests have been ignored. Reasons for the fork are:
 * fixes a white space bug
 * unittest based test suite
 * support for CPython and Jython  2.2 up to and including Python 3.x

I had a need for this to work with a VERY old JVM/JRE.

The things I like about hyperturtle's Stache is the exists check, more details in https://github.com/mustache/spec/discussions/171 which is essentially option 3 in the discussion at https://github.com/mustache/spec/discussions/157. This was basically my requirements when I (originally) decided to use Stache, exists-check and supporting almost any version of Python you can find.

Stache also has options for generating js/javascript templates, which is neat but it isn't something I've made use of. Additional options are documented in the readme (but some of those are now in the official Mustache spec now) https://github.com/clach04/Stache#new-stuff.

I'm not sure I'd recommend Stache to anyone starting now in 2023, look for something that isn't carrying around Python 2.2 baggage, but it works well for me ;-)

**Comment by jgonggrijp** [comment]
Nice nuanced story. Are there any features that you miss in Stache? If so, are you considering to add them?

**Reply by clach04** [reply]
1. add an option to print warning/error for template variables that are missing/not-passed-in (not just falsey)
2. add an option to print warning/error for variables that are passed in but missing in the template
3. support the official spec json/yaml files for easier specification compliance testing! :-)
    *  Stache is a library rather than a command line tool

I **think** briefly toyed with 1 and 2 (but I clearly wasn't that committed to it as I can't find the experimental code anymore!). Definitely useful for new users (or initial development/debugging).

I've definitely seen people in general asking about 1 and 2 (but no requests for the Stache implementation)
